src/components/MessageInput/hooks/useUserTrigger.ts

```diff
@@ -0,0 +1,158 @@
+  DefaultUserType,
+} from '../../../../types/types';
+import type { UserResponse } from 'stream-chat';
+
+const useUserTrigger = <
+  At extends DefaultAttachmentType = DefaultAttachmentType,
+  Ch extends DefaultChannelType = DefaultChannelType,
+  Co extends DefaultCommandType = DefaultCommandType,
+  Ev extends DefaultEventType = DefaultEventType,
+  Me extends DefaultMessageType = DefaultMessageType,
+  Re extends DefaultReactionType = DefaultReactionType,
+  Us extends DefaultUserType<Us> = DefaultUserType
+>(
+  mentionQueryParams: MentionQueryParams<Us> = {},
+  onSelectItem: (item: UserResponse<Us>) => void,
+  mentionAllAppUsers?: boolean,
+): UserTriggerSetting<Us> => {
+  const [searching, setSearching] = useState(false);
+
+  const { client } = useChatContext<At, Ch, Co, Ev, Me, Re, Us>();
+  const { channel } = useChannelStateContext<At, Ch, Co, Ev, Me, Re, Us>();
+  const members = channel?.state?.members;
+  const watchers = channel?.state?.watchers;
+
+  const getMembersAndWatchers = useCallback(() => {
+    const memberUsers = members ? Object.values(members).map(({ user }) => user) : [];
+    const watcherUsers = watchers ? Object.values(watchers) : [];
+    const users = [...memberUsers, ...watcherUsers];
+
+    // make sure we don't list users twice
+    const uniqueUsers = {} as Record<string, UserResponse<Us>>;
+
+    users.forEach((user) => {
+      if (user && !uniqueUsers[user.id]) {
+        uniqueUsers[user.id] = user;
+      }
+    });
+
+    return Object.values(uniqueUsers);
+  }, [members, watchers]);
+
+  const queryMembersDebounced = useCallback(
+    debounce(async (query: string, onReady: (users: UserResponse<Us>[]) => void) => {
+      try {
+        // @ts-expect-error
+        const response = await channel.queryMembers({
+          name: { $autocomplete: query },
+        });
+
+        const users = response.members.map((member) => member.user) as UserResponse<Us>[];
+
+        if (onReady && users.length) {
+          onReady(users);
+        } else {
+          onReady([]);
+        }
+      } catch (error) {
+        console.log({ error });
+      }
+    }, 200),
+    [channel],
+  );
+
+  const queryUsers = async (query: string, onReady: (users: UserResponse<Us>[]) => void) => {
+    if (!query || searching) return;
+    setSearching(true);
+
+    try {
+      const { users } = await client.queryUsers(
+        // @ts-expect-error
+        {
+          $or: [{ id: { $autocomplete: query } }, { name: { $autocomplete: query } }],
+          id: { $ne: client.userID },
+          ...mentionQueryParams.filters,
+        },
+        { id: 1, ...mentionQueryParams.sort },
+        { limit: 10, ...mentionQueryParams.options },
+      );
+
+      if (onReady && users.length) {
+        onReady(users);
+      } else {
+        onReady([]);
+      }
+    } catch (error) {
+      console.log({ error });
+    }
+
+    setSearching(false);
+  };
+
+  const queryUsersDebounced = debounce(queryUsers, 200);
```

Comment:
i change this to throttle recently after working on the channel search stuff, realized throttle was more appropriate for our use case, so might be a merge conflict since it was moved to a new file/hook